misc/python/materialize/cli/mzcompose.py

```diff
@@ -177,142 +189,251 @@ def gen_shortcuts(repo: mzbuild.Repository) -> int:
-        errors += mzcompose.Composition.lint(repo, name)
-    for error in sorted(errors):
-        print(error)
-    return 1 if errors else 0
-
-
-def list_compositions(repo: mzbuild.Repository) -> int:
-    for name in sorted(repo.compositions):
-        print(name)
-    return 0
-
-
-def list_workflows(composition: mzcompose.Composition) -> int:
-    for name in sorted(
-        list(composition.yaml_workflows) + list(composition.python_funcs)
-    ):
-        print(name)
-    return 0
-
-
-def list_ports(composition: mzcompose.Composition, service: Optional[str]) -> int:
-    if service is None:
-        raise errors.MzRuntimeError(f"list-ports command requires a service argument")
-    for port in composition.find_host_ports(service):
-        print(port)
-    return 0
-
-
-def web(composition: mzcompose.Composition, service: Optional[str]) -> int:
-    if service is None:
-        raise errors.MzRuntimeError(f"web command requires a service argument")
-    ports = composition.find_host_ports(service)
-    if len(ports) == 1:
-        webbrowser.open(f"http://localhost:{ports[0]}")
-    elif not ports:
-        raise errors.MzRuntimeError(f"No running services matched {service!r}")
-    else:
-        raise errors.MzRuntimeError(
-            f"Too many ports matched {service!r}, found: {ports}"
+        for path in repo.compositions.values():
+            mzcompose_path = path / "mzcompose"
+            with open(mzcompose_path, "w") as f:
+                f.write(template.format(os.path.relpath(repo.root, path)))
+            mzbuild.chmod_x(mzcompose_path)
+
+
+class LintCommand(Command):
+    name = "lint"
+    help = "surface common errors in compositions"
+
+    def run(cls, args: argparse.Namespace) -> None:
+        super().run(args)
+        repo = load_repo(args)
+        errors = []
+        for name in repo.compositions:
+            errors += mzcompose.Composition.lint(repo, name)
+        for error in sorted(errors):
+            print(error)
+        if errors:
+            sys.exit(1)
+
+
+class ListCompositionsCommand(Command):
+    name = "list-compositions"
+    help = "list the directories that contain compositions"
+
+    def run(cls, args: argparse.Namespace) -> None:
+        super().run(args)
+        repo = load_repo(args)
+        for name in sorted(repo.compositions):
+            print(name)
+
+
+class ListPortsCommand(Command):
+    name = "list-ports"
+    help = "list ports exposed by a service"
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            "service", metavar="SERVICE", help="the service to list ports for"
         )
-    return 0
+
+    def run(self, args: argparse.Namespace) -> None:
+        super().run(args)
+        composition = load_composition(args)
+        for port in composition.find_host_ports(args.service):
+            print(port)
+
+
+class ListWorkflowsCommand(Command):
+    name = "list-workflows"
```

Comment:
i'd make `list` a shorthand for this, `add_parser` takes `aliases=List[str]`

Reply:
That's a good idea. If you don't mind I'll address that in a follow-up PR though since I want to avoid making too many behavior changes in this mega PR.

Reply:
I like to add TODOs so I am less likely to forget!